model/atmosphere/dycore/src/icon4py/model/atmosphere/dycore/solve_nonhydro.py

```diff
@@ -436,127 +436,201 @@ def __init__(
-        )
+        ).compile(offset_provider_type=self._grid.offset_providers)
         self._predictor_stencils_11_lower_upper = (
-            nhsolve_stencils.predictor_stencils_11_lower_upper.with_backend(self._backend)
+            nhsolve_stencils.predictor_stencils_11_lower_upper.with_backend(self._backend).compile(
+                offset_provider_type=self._grid.offset_providers
+            )
         )
         self._predictor_stencils_35_36 = nhsolve_stencils.predictor_stencils_35_36.with_backend(
             self._backend
+        ).compile(offset_provider_type=self._grid.offset_providers)
+        self._predictor_stencils_37_38 = (
+            nhsolve_stencils.predictor_stencils_37_38.with_backend(self._backend)
+            .with_connectivities(self._grid.offset_providers)
+            .freeze()
         )
-        self._predictor_stencils_37_38 = nhsolve_stencils.predictor_stencils_37_38.with_backend(
-            self._backend
+        self._stencils_39_40 = nhsolve_stencils.stencils_39_40.with_backend(self._backend).compile(
+            offset_provider_type=self._grid.offset_providers
         )
-        self._stencils_39_40 = nhsolve_stencils.stencils_39_40.with_backend(self._backend)
         self._stencils_43_44_45_45b = nhsolve_stencils.stencils_43_44_45_45b.with_backend(
             self._backend
+        ).compile(offset_provider_type={})
+        self._stencils_47_48_49 = nhsolve_stencils.stencils_47_48_49.with_backend(
+            self._backend
+        ).compile(offset_provider_type=self._grid.offset_providers)
+        self._stencils_61_62 = nhsolve_stencils.stencils_61_62.with_backend(self._backend).compile(
+            offset_provider_type={}
         )
-        self._stencils_47_48_49 = nhsolve_stencils.stencils_47_48_49.with_backend(self._backend)
-        self._stencils_61_62 = nhsolve_stencils.stencils_61_62.with_backend(self._backend)
         self._en_smag_fac_for_zero_nshift = smagorinsky.en_smag_fac_for_zero_nshift.with_backend(
             self._backend
-        )
-        self._init_test_fields = nhsolve_stencils.init_test_fields.with_backend(self._backend)
+        ).compile(offset_provider_type={"Koff": dims.KDim})
+        self._init_test_fields = nhsolve_stencils.init_test_fields.with_backend(
+            self._backend
+        ).compile(offset_provider_type={})
         self._stencils_42_44_45_45b = nhsolve_stencils.stencils_42_44_45_45b.with_backend(
             self._backend
-        )
+        ).compile(offset_provider_type={})
+
         if self._config.divdamp_type == 32:
-            xp = data_alloc.import_array_ns(self.backend)
-            self.starting_vertical_index_for_3d_divdamp = xp.min(
-                xp.where(self._metric_state_nonhydro.scaling_factor_for_3d_divdamp.ndarray > 0.0)
-            )
+            xp = data_alloc.import_array_ns(self._backend)
+            tmp = xp.where(self._metric_state_nonhydro.scaling_factor_for_3d_divdamp.ndarray > 0.0)
```

Comment:
this where is missing 2 arguments? but didn't have time what the code should do...